compiler/rustc_mir_transform/src/check_redundant_transmutes.rs

```diff
@@ -0,0 +1,114 @@
+        let mut checker = RedundantTransmutesChecker { body, tcx };
+        checker.visit_body(body);
+    }
+}
+
+struct RedundantTransmutesChecker<'a, 'tcx> {
+    body: &'a Body<'tcx>,
+    tcx: TyCtxt<'tcx>,
+}
+
+impl<'a, 'tcx> RedundantTransmutesChecker<'a, 'tcx> {
+    /// This functions checks many things:
+    /// 1. if the source (`transmute::<$x, _>`) is `[u8; _]`, check if the output is a `{uif}xx`
+    /// 2. swap and do the same check.
+    /// 3. in the case of `char` → `u32` suggest `to_u32` and `from_u32_unchecked`
+    /// 4. `uxx` → `ixx` should be `as`
+    /// Returns the replacement.
+    fn is_redundant_transmute(
+        &self,
+        function: &Operand<'tcx>,
+        arg: String,
+        span: Span,
+    ) -> Option<errors::RedundantTransmute> {
+        let fn_sig = function.ty(self.body, self.tcx).fn_sig(self.tcx).skip_binder();
+        let [input] = fn_sig.inputs() else { return None };
+
+        // Checks if `x` is `[u8; _]`
+        let is_u8s = |x: Ty<'tcx>| matches!(x.kind(), Array(t, _) if *t.kind() == Uint(UintTy::U8));
+        // dont check the length; transmute does that for us.
+        if is_u8s(*input) && matches!(fn_sig.output().kind(), Uint(..) | Float(_) | Int(_)) {
+            // FIXME: get the whole expression out?
+            return Some(errors::RedundantTransmute {
+                sugg: format!("{}::from_ne_bytes({arg})", fn_sig.output()),
+                help: Some(
+                    "there's also `from_le_bytes` and `from_ne_bytes` if you expect a particular byte order",
+                ),
+                span,
+            });
+        }
+        if is_u8s(fn_sig.output()) && matches!(input.kind(), Uint(..) | Float(_) | Int(_)) {
+            return Some(errors::RedundantTransmute {
+                sugg: format!("{input}::to_ne_bytes({arg})"),
+                help: Some(
+                    "there's also `to_le_bytes` and `to_ne_bytes` if you expect a particular byte order",
+                ),
+                span,
+            });
+        }
+        return match input.kind() {
+            // char → u32
+            Char => matches!(fn_sig.output().kind(), Uint(UintTy::U32)).then(|| {
+                errors::RedundantTransmute { sugg: format!("({arg}) as u32"), help: None, span }
+            }),
+            // u32 → char
+            Uint(UintTy::U32) if *fn_sig.output().kind() == Char => {
+                Some(errors::RedundantTransmute {
+                    sugg: format!("char::from_u32_unchecked({arg})"),
+                    help: Some("consider `char::from_u32(…).unwrap()`"),
+                    span,
+                })
+            }
+            // bool → (uNN ↔ iNN)
+            Bool | Uint(..) | Int(..) => {
+                matches!(fn_sig.output().kind(), Int(..) | Uint(..)).then(|| {
+                    errors::RedundantTransmute {
+                        sugg: format!("({arg}) as {}", fn_sig.output()),
```

Comment:
I wouldn't suggest an as conversion here as as is so overloaded that I at least always forget what each conversion does. I remember there was an attempt in the past to create explicit methods for each as behaviour - if they've been merged, perhaps we could recommend them?

Reply:
im not sure what methods you're referring to... I dont think i should suggest `i32::from_ne_bytes(x.to_ne_bytes())`? And `try_from` has different behaviour.

Reply:
There are the unstable [`cast_unsigned`](https://doc.rust-lang.org/std/primitive.i32.html#method.cast_unsigned) and [`cast_signed`](https://doc.rust-lang.org/std/primitive.u32.html#method.cast_signed) methods, and there's an ACP for additional methods https://github.com/rust-lang/libs-team/issues/453.

Perhaps for now the code just gets a FIXME so that the recommendation is changed once https://github.com/rust-lang/rust/issues/125882 is stabilized (feature `integer_sign_cast`)?

Reply:
oh wow! i guess ill make it suggest those methods then

Reply:
what about `bool as integer` though?

Reply:
@bend-n
```rust
    let b: bool = true;
    let x = u8::from(b);
    assert_eq!(x, 1);
```

`.into()` would be nicer but depends on knowing `u8` or `i8` from type inference. `u8::from` and `i8::from` work in any context, though they don't work in postfix position. For the purposes of *this* lint, though, the only integer types you can `transmute` a `bool` to would be `u8` or `i8`, and `transmute` already isn't postfix, so I'd suggest that transmutes from `bool` to `u8`/`i8` should suggest `u8::from` and `i8::from`, respectively.

Reply:
i can do that with `u32::from(char)` i think, as well